File: rxbinding-support-v4/src/main/java/com/jakewharton/rxbinding2/support/v4/view/ViewPagerPageScrolledObservable.java
package com.jakewharton.rxbinding2.support.v4.view;

import android.support.v4.view.ViewPager;
import android.support.v4.view.ViewPager.OnPageChangeListener;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class ViewPagerPageScrolledObservable extends Observable<ViewPagerPageScrollEvent> {
  private final ViewPager view;

  ViewPagerPageScrolledObservable(ViewPager view) {
    this.view = view;
  }

  @Override
  protected void subscribeActual(Observer<? super ViewPagerPageScrollEvent> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnPageChangeListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnPageChangeListener {
    private final ViewPager view;
    private final Observer<? super ViewPagerPageScrollEvent> observer;

    Listener(ViewPager view, Observer<? super ViewPagerPageScrollEvent> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
      if (!isDisposed()) {
        ViewPagerPageScrollEvent event =
            ViewPagerPageScrollEvent.create(view, position, positionOffset,
                positionOffsetPixels);
        observer.onNext(event);
      }
    }

    @Override
    public void onPageSelected(int position) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    @Override
    protected void onDispose() {
      view.removeOnPageChangeListener(this);
    }
  }
}

File: rxbinding-viewpager/src/main/java/com/jakewharton/rxbinding3/viewpager/ViewPagerPageScrolledObservable.kt
@file:JvmName("RxViewPager")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.viewpager

import androidx.annotation.CheckResult
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding2.internal.checkMainThread

/**
 * Create an observable of page scroll events on `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun ViewPager.pageScrollEvents(): Observable<ViewPagerPageScrollEvent> {
  return ViewPagerPageScrolledObservable(this)
}

data class ViewPagerPageScrollEvent(
  val viewPager: ViewPager,
  val position: Int,
  val positionOffset: Float,
  val positionOffsetPixels: Int
)

private class ViewPagerPageScrolledObservable(
  private val view: ViewPager
) : Observable<ViewPagerPageScrollEvent>() {

  override fun subscribeActual(observer: Observer<in ViewPagerPageScrollEvent>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addOnPageChangeListener(listener)
  }

  private class Listener(
    private val view: ViewPager,
    private val observer: Observer<in ViewPagerPageScrollEvent>
  ) : MainThreadDisposable(), OnPageChangeListener {

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
      if (!isDisposed) {
        val event = ViewPagerPageScrollEvent(view, position, positionOffset,
            positionOffsetPixels)
        observer.onNext(event)
      }
    }

    override fun onPageSelected(position: Int) {

    }

    override fun onPageScrollStateChanged(state: Int) {

    }

    override fun onDispose() {
      view.removeOnPageChangeListener(this)
    }
  }
}
